### DISCUSSION TITLE
FluxCD Deployments removed reinstall deleted everything, is that normal?

### DISCUSSION THREAD

**Discussion by qvistgaard** [fluxcd/flux2#4094] [Q&A] [answered]
Hi

So we have a test cluster, this morning I could not understand why FluxCD was not applying any changes. So I start debugging and see some warnings about git repo sources not existing. No biggie, I go to check if something is up with the FluxCD pods, and they are gone, as the deployments are gone but the namespace is still there. How they disappeared, does not matter.

Next, No biggie, I'll just reinstall FluxCD using terraform. I run the terraform, and excitedly go to the cluster to watch in awe as flux restores the state and ..... *SURPRISED PIKACHU FACE* .. It started deleting everything.  including the flux-system namespace

So my question is What the f just happened? Any one got any ideas suggestion to why fluxcd would do such a thing?

**Comment by kingdonb** [comment]
Are you using the old flux_install terraform provider or the new terraform-provider-flux with flux_bootstrap_git provider resource?

I think what you're describing sounds like an issue I've been hearing about with the old provider, but there's not really enough detail to confirm it without looking at your terraform files. I've been hearing that the kubernetes provider can delete CRDs or namespaces during an upgrade if you don't read the plan carefully. Deleting the CRD from an active Flux installation could possibly trigger garbage collection, which could explain the deletion of your Flux-system namespace. I'm fairly sure that Flux isn't responsible but I haven't seen a full repro, so I can't say this with certainty. I'll tell about how some of the components work and hopefully this sheds some light.

The old Flux terraform provider cannot delete your resources, it only emits YAML. The Kubernetes provider is the provider from Terraform that is used in the old docs, and it could delete them. There should be no way that bootstrapping with the new Flux provider (flux_bootstrap_git) would delete the Flux-system namespace or your CRDs, since it uses the actual Flux bootstrap code from upstream Flux packages, that is designed to be idempotent and apply over itself / only apply and never delete resources.

If your terraform resources for Flux are over about 6 months old, you are using the old resource. It's completely different from the new resource, the new one is much simpler and does not require so many external/other providers. When you upgrade to Flux 2.0.0 (via the Terraform provider 1.0.0 and above) you will see a CRD upgrade, and I cannot confirm the old Kubernetes provider for Terraform will handle the CRD upgrade properly. That could be the whole source of the issue if you did not switch to migrate to the new resource.

https://github.com/fluxcd/terraform-provider-flux#get-started

There is a migration guide to help with the finer points as you upgrade from the old resource provider kind to the new provider:

https://github.com/fluxcd/terraform-provider-flux/blob/main/docs/guides/migrating-to-resource.md

**Reply by qvistgaard** [reply]
We are actually using the the old version, we just never got around to updating it yet. but would that also explain why deployments and namespaces managed by flux would be deleted?

**Reply by kingdonb** [reply]
If the flux CRDs are deleted, while Flux controllers are running, that would send all the CR instances into Terminating, which would trigger garbage collection. This would only happen if the flux controllers were running and `prune` was enabled on the Kustomization. I have not analyzed the behavior of the Kubernetes provider to say for sure exactly what is happening.

Here is a related thread where someone else experienced the similar issue, they were also using the old flux_install provider:

https://github.com/fluxcd/terraform-provider-flux/issues/523

**Reply by stefanprodan** [reply, marked as answer]
The old provider had many issues, making Flux delate itself via Kubernetes GC was one of it. Please upgrade to the new provider.
